masonry/src/passes/update.rs

```diff
@@ -0,0 +1,163 @@
+            .find_widget_at_pos(pos)
+            .map(|widget| widget.id())
+    } else {
+        None
+    };
+    if let Some(capture_target) = root.state.pointer_capture_target {
+        if next_hovered_widget != Some(capture_target) {
+            next_hovered_widget = None;
+        }
+    }
+
+    let prev_hovered_path = std::mem::take(&mut root.state.hovered_path);
+    let next_hovered_path = get_id_path(root, next_hovered_widget);
+
+    let mut hovered_set = HashSet::new();
+    for widget_id in &next_hovered_path {
+        hovered_set.insert(*widget_id);
+    }
+
+    trace!("prev_hovered_path: {:?}", prev_hovered_path);
+    trace!("next_hovered_path: {:?}", next_hovered_path);
+
+    // This algorithm is written to be resilient to future changes like reparenting and multiple
+    // cursors. In theory it's O(Depth² * CursorCount) in the worst case, which isn't too bad
+    // (cursor count is usually 1 or 2, depth is usually small), but in practice it's virtually
+    // always O(Depth * CursorCount) because we only need to update the hovered status of the
+    // widgets that changed.
+    // The above assumes that accessing the widget tree is O(1) for simplicity.
+
+    fn update_hovered_status_of(
+        root: &mut RenderRoot,
+        widget_id: WidgetId,
+        hovered_set: &HashSet<WidgetId>,
+    ) {
+        run_targeted_update_pass(root, Some(widget_id), |widget, ctx| {
+            let is_hot = hovered_set.contains(&ctx.widget_id());
+
+            if ctx.widget_state.is_hot != is_hot {
+                widget.on_status_change(ctx, &StatusChange::HotChanged(is_hot));
+            }
+            ctx.widget_state.is_hot = is_hot;
+        });
+    }
+
+    // TODO - Make sure widgets are iterated from the bottom up
+    for widget_id in &prev_hovered_path {
+        if root
+            .widget_arena
+            .widget_states
+            .find(widget_id.to_raw())
+            .is_some()
+            && root.widget_arena.get_state_mut(*widget_id).item.is_hot
+                != hovered_set.contains(widget_id)
+        {
+            update_hovered_status_of(root, *widget_id, &hovered_set);
+        }
+    }
+    for widget_id in &next_hovered_path {
+        if root
+            .widget_arena
+            .widget_states
+            .find(widget_id.to_raw())
+            .is_some()
+            && root.widget_arena.get_state_mut(*widget_id).item.is_hot
+                != hovered_set.contains(widget_id)
+        {
+            update_hovered_status_of(root, *widget_id, &hovered_set);
+        }
+    }
+
+    root.state.hovered_path = next_hovered_path;
+
+    // -- UPDATE CURSOR --
+    // TODO - Rewrite more cleanly
+    let cursor_changed = next_hovered_widget
+        .is_some_and(|id| root.widget_arena.get_state_mut(id).item.cursor_changed);
+    if hovered_widget != next_hovered_widget || cursor_changed {
+        let cursor;
+        if let Some(capture_target) = root.state.pointer_capture_target {
```

Comment:
I'm presuming the case where the capture target's `cursor_changed` is set is handled elsewhere.

Reply:
... Crap, I don't think it is. Good catch.

Reply:
Re-reading this, I think I was assuming `capture_target` would always be the same as `hovered_widget`, but that's not actually the case (`hovered_widget` could be None it we'd still want to check the capture target).

In the end I just removed the `cursor_changed` part for expediency.